### ISSUE TITLE
”ValueError: No objects to concatenate“ while computing features.

### ISSUE THREAD

**Issue by Visytudz** [AllenCell/cvapipe_analysis#53]
Hello! I am reaching out to report an issue I've encountered while attempting to reproduce the project that I previously managed to run successfully. I noticed that the process now fails during the second step, which involves computing features, and the error seems to originate from the `load_results_in_single_dataframe()` function, though I am not certain.

I have checked out to the nature_paper branch and have only made modifications to the storage paths in the config.yaml file to suit my local environment, with no other changes made.
```
[INFO: 196 2024-03-22 23:18:38,324] Attempted file (/root/project/reproduce/loaddata/crop_seg/e6af8064dbf16d5a684c45583248e77084f6b1fe7d74780cfa7b54e8ddf4f594_segmentation.ome.tif) load with reader: aicsimageio.readers.bfio_reader.OmeTiledTiffReaderfor better performance, consider installing 'bfio'.
[INFO: 196 2024-03-22 23:18:38,328] Attempted file (/root/project/reproduce/loaddata/crop_seg/c91fc147996b0c01756df9c17bcea9307f1bf33105afcb4cf5bb321db4d74300_segmentation.ome.tif) load with reader: aicsimageio.readers.bfio_reader.OmeTiledTiffReaderfor better performance, consider installing 'bfio'.
[INFO:  48 2024-03-22 23:19:03,552] Loading results...
0it [00:00, ?it/s]
Traceback (most recent call last):
  File "/root/miniconda3/envs/cvapipe/bin/cvapipe_analysis", line 33, in <module>
    sys.exit(load_entry_point('cvapipe-analysis', 'console_scripts', 'cvapipe_analysis')())
  File "/root/project/cvapipe_analysis/cvapipe_analysis/bin/cli.py", line 35, in cli
    fire.Fire({**step_map, "all": All})
  File "/root/miniconda3/envs/cvapipe/lib/python3.8/site-packages/fire/core.py", line 143, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/root/miniconda3/envs/cvapipe/lib/python3.8/site-packages/fire/core.py", line 477, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/root/miniconda3/envs/cvapipe/lib/python3.8/site-packages/fire/core.py", line 693, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/root/miniconda3/envs/cvapipe/lib/python3.8/site-packages/datastep/step.py", line 60, in wrapper
    return func(self, *args, **kwargs)
  File "/root/project/cvapipe_analysis/cvapipe_analysis/steps/compute_features/compute_features.py", line 49, in run
    df_results = calculator.load_results_in_single_dataframe()
  File "/root/project/cvapipe_analysis/cvapipe_analysis/tools/io.py", line 190, in load_results_in_single_dataframe
    df = pd.concat(
  File "/root/miniconda3/envs/cvapipe/lib/python3.8/site-packages/pandas/util/_decorators.py", line 331, in wrapper
    return func(*args, **kwargs)
  File "/root/miniconda3/envs/cvapipe/lib/python3.8/site-packages/pandas/core/reshape/concat.py", line 368, in concat
    op = _Concatenator(
  File "/root/miniconda3/envs/cvapipe/lib/python3.8/site-packages/pandas/core/reshape/concat.py", line 425, in __init__
    raise ValueError("No objects to concatenate")
ValueError: No objects to concatenate
```
Could you please advise on how I might resolve this issue? If additional information or logs are required for diagnosis, please let me know, and I will provide them promptly.

Thank you very much for your assistance and support!

**Comment by vianamp** [comment]
Could you please post your config file and a list of directories you have in your local staging?

**Comment by Visytudz** [comment]
```
/root/project/
├── cvapipe_analysis
│   ├── codecov.yml
│   ├── CODE_OF_CONDUCT.md
│   ├── config.yaml
│   ├── CONTRIBUTING.md
│   ├── cvapipe_analysis
│   ├── cvapipe_analysis.egg-info
│   ├── docs
│   ├── LICENSE
│   ├── Makefile
│   ├── MANIFEST.in
│   ├── notebooks
│   ├── README.md
│   ├── setup.cfg
│   ├── setup.py
│   ├── tox.ini
│   └── workflow_config.json
├── reproduce
│   ├── computefeatures
│   ├── loaddata
│   └── manifest.csv
└── scripts
    └── file_counter.sh

reproduce/
├── computefeatures
│   ├── cell_features
│   ├── init_parameters.json
│   ├── parameters.yaml
│   └── run_parameters.json
├── loaddata
│   ├── crop_raw
│   ├── crop_seg
│   ├── init_parameters.json
│   ├── manifest.csv
│   ├── parameters.yaml
│   └── run_parameters.json
└── manifest.csv
```
[config.txt](https://github.com/AllenCell/cvapipe_analysis/files/14725109/config.txt)
I changed the .yaml to .txt because the former couldn't be uploaded to issues.

**Comment by vianamp** [comment]
And you are running the test example? Or running it with your own dataset? Also, which version of Python are you using?

**Comment by Visytudz** [comment]
I use the test sample, i.e. 300 pairs of images provided at random. My python version is 3.8, which is the same as in the readme.

**Comment by vianamp** [comment]
OK. Thank you for the information. I am trying to reproduce it on my end. More soon.

**Comment by vianamp** [comment]
Hi @Visytudz. I was able to run the code with no problem. This is my workflow:
```
conda create --name cva python=3.8
git clone git@github.com:AllenCell/cvapipe_analysis.git --branch nature-paper --single-branch
cd cvapipe_analysis
pip install -e .
```
Edit the config file with your local staging path.
```
cvapipe_analysis load_data run --test
cvapipe_analysis computefeatures run
```
Final output:
```
[INFO:  48 2024-03-22 11:38:34,985] Loading results...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 300/300 [00:01<00:00, 250.00it/s]
[INFO:  59 2024-03-22 11:38:36,886] Saving manifest...
```
Please let me know if this still does not work for you.

**Comment by Visytudz** [comment]
Hi @vianamp. It appears that I have discovered the source of the problem. Essentially, the issue arises from not waiting for the asynchronous parallel feature computations to finish before proceeding with the rest of the program. 

In your case, with more robust computing resources, the calculations might have completed before the program advanced (though I find the speed quite surprising). However, on my less powerful machine, the computations did not complete by the time the program attempted to execute `df_results = calculator.load_results_in_single_dataframe()`—it seemed as if none of the data had been processed. Consequently, the directory `reproduce/computefeatures/cell_features/` was empty, resulting in the error.

To address this, I initially inserted a statement to await the end of the asynchronous operations, but encountered an error indicating that my CPU was at its limit. Subsequently, I adjusted the number of concurrent cores to `control.get_ncores() // 2`, which successfully mitigated the issue.
![image](https://github.com/AllenCell/cvapipe_analysis/assets/83176441/26b254d3-59b2-452a-a63a-25e1b693d352)


As a newcomer to the fields of Python programming and data analysis, navigating this challenge has been quite daunting! I really appreciate your patience and assistance with this issue, which rarely encountered in daily interactions. Should you have a more refined solution, I am eager to learn, and I am hopeful that this insight will contribute to the project's ongoing refactoring efforts.
